### PR TITLE
DO UPDATE support for backfill operations

### PR DESCRIPTION
**Pull request by noctarius**
The current backfill script only supports errors on duplicate rows or a _DO NOTHING_ clause. This pull request provides a first draft of general support for column updates in backfill operations.

Columns to update can be passed to the function using the new parameter update_action_columns as `array['column1', 'column2', ...]` and are used to generate the actual DO UPDATE SET clause when on_conflict_action is set to 'Update'.

### CODE REVIEW

**Comment by noctarius** [comment]
One thing that is missing to make it work (and there is probably some internal hidden magic), what would be the "on conflict constraint"? Is there some internal thing, as there is no default PK.

File: backfill.sql [hunk 1]
```diff
@@ -29,6 +29,10 @@
 * of that table in certain ranges faster. (ie `CREATE INDEX ON cpu_temp(time);`)
 */
 
+---- Backfill operation enum type to define the behavior in case of conflicting rows
+DROP TYPE IF EXISTS BackfillOnConflict;
+CREATE TYPE BackfillOnConflict AS ENUM ('DoNothing', 'Update', 'Error');
```
Comment (davidkohn88) [inline backfill.sql:34]: A couple comments: 
1) I think I'd have the names here be 'NOTHING', 'UPDATE', 'RESTRICT' just to keep them consistent with PG naming.
2) I'm not sure if I'd actually use an ENUM here or if I'd just pass in text to the function and then compare it there and error if it's not one of the three, mainly because I think ENUMs are case sensitive, (so I'd actually compare UPPER(foo) or whatever) and I think it'd be weird to have to match the case so closely.

File: backfill.sql [hunk 2]
```diff
@@ -181,8 +189,17 @@ BEGIN
...
     END IF;
 
-    IF on_conflict_do_nothing THEN
+    IF on_conflict_action = 'DoNothing' THEN
         on_conflict_clause = 'ON CONFLICT DO NOTHING';
+    ELSEIF on_conflict_action = 'Update' THEN
+        on_conflict_clause = 'ON CONFLICT DO UPDATE SET ';
+        FOR on_conflict_index IN 1 .. ARRAY_LENGTH(update_action_columns, 1) LOOP
+            on_conflict_column = update_action_columns[on_conflict_index];
+            on_conflict_clause = on_conflict_clause || FORMAT('%2$s.%1$I = to_insert.%1$I', on_conflict_column, dest);
```
Comment (davidkohn88) [inline backfill.sql:198]: I think the format clause will be simpler if you do something like `FORMAT('%1$I = EXCLUDED.%1$I', on_conflict_column)`

File: backfill.sql [hunk 3]
```diff
@@ -181,8 +189,17 @@ BEGIN
             $$;
     END IF;
 
-    IF on_conflict_do_nothing THEN
+    IF on_conflict_action = 'DoNothing' THEN
         on_conflict_clause = 'ON CONFLICT DO NOTHING';
+    ELSEIF on_conflict_action = 'Update' THEN
+        on_conflict_clause = 'ON CONFLICT DO UPDATE SET ';
+        FOR on_conflict_index IN 1 .. ARRAY_LENGTH(update_action_columns, 1) LOOP
```
Comment (davidkohn88) [inline backfill.sql:196]: I would just do this with `SELECT 'ON CONFLICT DO UPDATE SET ' || STRING_AGG(FORMAT('%1$I = EXCLUDED.%1$I', on_conflict_column), ', ')`

File: backfill.sql [hunk 4]
```diff
@@ -97,14 +101,16 @@ $$ LANGUAGE PLPGSQL VOLATILE;
 -- The main event
 -- staging_table is the (possibly temporary) table from which rows will be moved 
 -- destination_hypertable is the table where rows will be moved
--- on_conflict_do_nothing controls whether duplicate rows are ignored
+-- on_conflict_action controls whether duplicate rows are being ignored (DoNothing, default), an update clause is generated based on on_conflict_columns (Update), or if the duplicate row raises an error (Error) 
 -- delete_from_staging specifies whether we should delete from the staging table as we go or leave rows there
 -- compression_job_push_interval specifies how long push out the compression job as we are running, ie the max amount of time you expect the backfill to take
+-- update_action_columns is an array of columns to use in the update clause when a conflict arises and the on_conflict_action is set to 'Update'
```
Comment (davidkohn88) [inline backfill.sql:107]: I would probably name this: `on_conflict_update_columns`

File: backfill.sql [hunk 5]
```diff
@@ -97,14 +101,16 @@ $$ LANGUAGE PLPGSQL VOLATILE;
 -- The main event
 -- staging_table is the (possibly temporary) table from which rows will be moved 
 -- destination_hypertable is the table where rows will be moved
--- on_conflict_do_nothing controls whether duplicate rows are ignored
+-- on_conflict_action controls whether duplicate rows are being ignored (DoNothing, default), an update clause is generated based on on_conflict_columns (Update), or if the duplicate row raises an error (Error)
```
Comment (davidkohn88) [inline backfill.sql:104]: This probably needs to be reworded, it's gotten a little long and hard to follow:
on_conflict action controls how duplicate rows are handled on insert, it has 3 allowed values that correspond to the ON CONFLICT actions in Postgres, 'NOTHING' (default), 'UPDATE', 'RESTRICT'. Their actions:
- NOTHING: ignore conflicting rows, the first inserted takes precedence
- UPDATE: replace values in the conflicting row according to the `on_conflict_update_columns` parameter, if this is set, the `on_conflict_update_columns` parameter must be set
- RESTRICT: error if there is a conflicting insert.

File: backfill.sql (hunk 4 above)
Comment (davidkohn88) [reply backfill.sql:107]: I'd also add a quick error checking step to check that this is set IFF on_conflict_do='UPDATE' and throw an error saying so.

**Review by davidkohn88 — COMMENTED** [review]
This is great! Thanks! I added some comments and it'd be great if you can manually test in a few scenarios what happens, especially if there's a misspelled column name etc.

**Comment by noctarius** [comment]
> Yeah. that's something to think about, but I think I would make it a separate PR and it might be a bit hairy to get to work so, we can leave that for now.

The problem is without that it'll error out if no PK is available. Not sure it shouldn't be at least handled by querying the pk or unique index definitions. WDYT?

**Comment by noctarius** [comment]
Updated code based on comments

**Review by davidkohn88 — APPROVED** [review]
Looks good to me! Sorry I missed this! Feel free to re-request a review in future, I'll be more likely to see it!
